src/bsk201910/rollbacks.md

```diff
@@ -0,0 +1,407 @@
+```
+
+So, what happens when you find a regression in the current release and decide to rollback to version 2 like so:
+
+```terminal|command=1|title=bash
+kubectl rollout undo deployment/app --to-revision=2
+```
+
+- Kubectl finds the ReplicaSets that belong to the Deployment
+- Each ReplicaSet has a revision number. Revision 2 is selected
+- The current replicas count is decreased, and the count is gradually increased in the RepiicaSet belonging to revision 2
+- The `deployment.kubernetes.io/revision` annotation is updated. The current ReplicaSet changes from revision 2 to 4
+
+If before the undo you had three ReplicaSets with revision 1, 2 and 3, now you should have 1, 3 and 4.
+
+There's a missing entry in the history: the revision 2 that was promoted to 4.
+
+_There's also something else that looks useful but doesn't work quite right._
+
+The history command displays two columns: _Revision_ and _Change-Cause_.
+
+```terminal|command=1|title=bash
+kubectl rollout history deployment/app
+REVISION  CHANGE-CAUSE
+1         <none>
+2         <none>
+3         <none>
+```
+
+While you're now familiar with the Revision column, you might be wondering what Change-Cause is used for — and why it's always set to `<none>`.
+
+When you create a resource in Kubernetes, you can append the `--record` flag like so:
+
+```terminal|command=1|title=bash
+kubectl create -f deployment.yaml --record
+```
+
+When you do, Kubernetes adds an annotation to the resource with the command that generated it.
+
+In the example above, the Deployment has the following metadata section:
+
+```yaml|highlight=5|title=deployment.yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: example-deployment
+  kubernetes.io/change-cause: kubectl create --filename=deployment.yaml
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      name: app
+  template:
+    metadata:
+      labels:
+        name: app
+    spec:
+      containers:
+      - name: app
+        image: learnk8s/hello:1.0.0
+```
+
+Now, if you try to display the history again, you might notice that the same annotation is used in the rollout history command:
+
+```terminal|command=1|title=bash
+kubectl rollout history app
+REVISION  CHANGE-CAUSE
+1         kubectl create --filename=deployment.yaml --record=true
+```
+
+If you change the container image in the YAML file and apply the new configuration with:
+
+```terminal|command=1|title=bash
+kubectl apply -f deployment.yaml --record
+```
+
+You should see the following new entry in the rollout history:
+
+```terminal|command=1|title=bash
+kubectl rollout history deployment/app
+REVISION  CHANGE-CAUSE
+1         kubectl create --filename=deployment.yaml --record=true
+2         kubectl apply --filename=deployment.yaml --record=true
+```
+
+The `--record` command can be used with any resource type, but it has an effect only on Deployment, DaemonSet, and StatefulSet resources, i.e. resources that can be "rolled out" (see `kubectl rollout -h`).
```

Comment:
I just tested it with `Service` and I think it works for everything, it's just an annotation (with the name of `kubernetes.io/change-cause`).

Reply:
Ahh, I think I understand your confusion now: the `--record` only has an effect for your object types, in the sense that only they have the rollout table.  But actually `--record` does have an effect for everything: putting in the annotation.  For other resource types, there is no nice command to get the annotation (rollout history), but you can still have it with `get -o yaml`.

Reply:
Weird isn't it :)

Reply:
Then maybe say, that "The --record command can be used with any resource type ...., but the recorded information can only be listed nicely for ...".  Because the "has an effect" is not true, it has an effect for everything (recording the annotation).  The fact that you want a nice list of the annotations is more related to rollout history than to --record.